### ISSUE TITLE
About multi-action output

### ISSUE THREAD

**Issue by dongxue1998** [ku2482/sac-discrete.pytorch#16]
How do I modify the code to fit the multi-action output？Will you upload multi-action output code later? @ku2482

**Comment by toshikwa** [comment]
@dongxue1998 
> How do I modify the code to fit the multi-action output?

The simplest way is to define the action space as the product of each action space.
Imagine that the environment has two action spaces, {left, None, right} and {forward, None, backward}.
In this case, we can define the action space which has 3*3=9 actions, therefore we can use this action space as ever.

> Will you upload multi-action output code later?

I don't have time for this now, however, I can help you to implement it and any pull requests are welcome.

Thanks :)

**Comment by dongxue1998** [comment]
Thank you for your reply.


I understand what you mean, but I want to know if I have 30 actions, and each action can selecte a number from 0-10, and if I modify code according to your way, my action space is too large, it will be very difficult to train,is there any way to solve this problem？





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ku2482/sac-discrete.pytorch"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年11月25日(星期三) 下午5:11
收件人:&nbsp;"ku2482/sac-discrete.pytorch"<sac-discrete.pytorch@noreply.github.com&gt;;
抄送:&nbsp;"Dye"<1521153032@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [ku2482/sac-discrete.pytorch] About multi-action output (#16)





 
@dongxue1998
  
How do I modify the code to fit the multi-action output?
  
The simplest way is to define the action space as the product of each action space.
 Imagine that the environment has two action spaces, {left, None, right} and {forward, None, backward}.
 In this case, we can define the action space which has 3*3=9 actions, therefore we can use this action space as ever.
  
Will you upload multi-action output code later?
  
I don't have time for this now, however, I can help you to implement it and any pull requests are welcome.
 
Thanks :)

**Comment by toshikwa** [comment]
@dongxue1998 

You can model the policy as multiple categorical distributions that share some layers.
\pi(a|s) = \pi_1(a_1|s) * \pi_2(a_2|s) * ...

However, we have to evaluate Q(s, a) = Q(s, a_1, a_2, ...), which is troublesome.
If we use a neural network that takes 's' as input and outputs scalars Q(s, a) for each action (in the same way as SAC-Discrete), the output dimension is too large.

One possible way is to model Q(s, a) as a neural network that takes 's' and (maybe one-hotted or normalized) 'a_i' as input and outputs a scalar. In this case, you can compute the expectation as the batch mean, not explicit calculation '\sum \pi(.|s) Q(s, .)'. It is no longer SAC-Discrete, it's rather Soft Actor-Critic.

I'm sorry that I can't come up with a smart solution.
Thanks.
